include/revng/Model/CommonTypeMethods.h

```diff
@@ -0,0 +1,171 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/ADT/RecursiveCoroutine.h"
+#include "revng/Model/PrimitiveKind.h"
+
+namespace model {
+class VerifyHelper;
+
+class TypeDefinition;
+class StructDefinition;
+
+class Type;
+class ArrayType;
+class PointerType;
+class PrimitiveType;
+
+template<typename CRTP>
+class CommonTypeMethods {
```

Comment:
As far as I intended, everything in here skips typedefs, but I might have missed something. I'll double-check tomorrow.

Reply:
Ah, then I need to take another look.

Reply:
I think these methods do too much automatically, this is not super-expected for the user.

1. The methods skip over typedefs, but the method name does not suggested that. These seem like upcasts to me, using them I wouldn't expect them to do any visit.
   I think it would be better to have a `.skipTypedefs` method and then you call `asPrimitive`, which doesn't do any visit. Having around a pattern such as `MyType.skipTypedefs().asPointer()` is more verbose but much clearer than `.asPointer()` doing magic.
2. As for `.skipTypedefs()` (or, in general, for method skipping typedefs), what do we do with `const`? We're inducing the user to ignore constness hidden into `typedef`s. This is one more reaons to make the skipping of `typedef` very explicit, have a single function for doing it and putting a big note stating that you're going to ignore `const` hidden in `typedef`s.
   Maybe we could have `skipTypedefs` and `resolveTypedefs`. The latter produces a new `model::Type` instance indentical to the the last `typedef`'d type but adding `IsConst = true` if any `typedef` in the `typedef` chain had `IsConst = true`.

What do you think?